ath79: specify N and ND subversions of TL-WR941 with ALT0_MODEL

adschm opened the pull request:
TP-Link's TL-WR941 is sold with detachable antennas
internationally (ND version), but with fixed antennas in China
(N version). While hardware and images are similar for both
variants of v2 and v4, they are different for v6.

Having both explicitly will make it easier for user to identify
the correct image, and mostly importantly will raise awareness
that N and ND are not necessarily always the same as for
TL-WR841 series.

With an image selection webpage, using ALT0_MODEL as in this
patch will provide the option to list versions for N and ND
separately.

~~This depends on PR #2250~~ 2250 is merged.

adschm committed:
ath79: specify N and ND subversions of TL-WR941 with ALT0_MODEL

TP-Link's TL-WR941 is sold with detachable antennas
internationally (ND version), but with fixed antennas in China
(N version). While hardware and images are similar for both
variants of v2 and v4, they are different for v6.

Having both explicitly will make it easier for user to identify
the correct image, and mostly importantly will raise awareness
that N and ND are not necessarily always the same as for
TL-WR841 series.

With an image selection webpage, using ALT0_MODEL as in this
patch will provide the option to list versions for N and ND
separately.

Signed-off-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>

adrianschmutzler commented:
@CodeFetch @981213 @aparcar

One might do it as in this patch OR:
If 941N versions are really only marketed in China, one might also really rename international devices to
tplink_tl-wr941nd-v2
and then introduced a cloned
tplink_tl-wr941n-v2-cn
to also indicate the difference as suggested by @981213 (but I'm not sure whether this was aiming at v2/v4).

CodeFetch commented:
@adrianschmutzler Yes, please rename the Chinese versions to "-cn". They don't have an FCC or CE logo and are thus not allowed to be imported to Europe, USA, New Zealand or Australia.

adrianschmutzler commented:
Well, I actually also tend to go into this direction, too, but it will cause lots of useless additional computation to build the same images twice. And to do it the proper way, one would have to backport the (albeit few) renames to 19.07. I personally think this would still be acceptable here, but I'd be interested in what the others are thinking.

981213 commented:
> (but I'm not sure whether this was aiming at v2/v4)

It aims at v6/v7 actually.
I think we don't need this for v2-v4. For v2-v4 they are actually the same board/firmware with no difference at all.

BTW: I just flashed the international WR941N firmware and here's the firmware version strings:
```
Firmware Version: | 3.13.9 Build 120201 Rel.54965n
Hardware Version: | WR940N v1/WR941N v4 00000000
```
TP-Link called it WR941N instead of ND in their firmware.

adrianschmutzler commented:
@981213 Okay, so I end up another time with trying to get order into something which nobody intended to keep ordered in the first place.

Thanks for looking that up, so I'm inclined to use this patch as it is right now without having separate device definitions.

CodeFetch commented:
Hm... I'm not really happy with this. Why do you use "v2/v3"? Why don't you add ALT devices for the versions, too? Aren't those version strings used for image generation? Even if it takes some disk space, a separate image should be provided for each device, shouldn't it?

adrianschmutzler commented:
@CodeFetch The DEVICE_MODEL etc. give the device names when using "make menuconfig" to select which devices should be built. The image names are constructed from the name of the device definition block in tiny-tp-link.mk (for ath79, ar71xx was different).
So, the user just taking images compiled by someone else will never see the DEVICE_(ALT_)MODEL, but only the image name. (Which is nevertheless worth caring about ...)

The problem of having several similar images is not so much space, but computation power. Especially if CONFIG_TARGET_PER_DEVICE_ROOTFS is enabled, a lot of computation will be redone several times to assemble images which are already there (you could just simply rename them).

This consumes some time and, more importantly, a lot of energy for the additional build. This will happen for the build machines as well as for every user just selecting the target and building every image no matter whether he needs it.
Thus, adding images is something you should have a reason for.

adrianschmutzler commented on target/linux/ath79/image/tiny-tp-link.mk:
```diff
@@ -183,8 +183,10 @@ TARGET_DEVICES += tplink_tl-wr841-v11
+  DEVICE_ALT0_MODEL := TL-WR941N
+  DEVICE_ALT0_VARIANT := v2/v3
   TPLINK_HWID := 0x09410002
   TPLINK_HWREV := 2
```
BTW: Is this needed? It's the only device in tiny-tp-link.mk having it defined.

981213 replied:
It's required. Otherwise TP-Link's OEM firmware will reject it.

CodeFetch commented:
Thank you for the clarification. In Gluon the model/board name is used for image naming and a separate image is created for each device and the identical images are symlinks. Maybe it would make sense to introduce this in OpenWrt at some point as it makes it easier for people to find the correct image for their device (so that they don't need to read the wiki)...

adschm committed:
ath79: Add separate device definition for TP-Link TL-WR941N/ND v3

TP-Link TL-WR94x devices are a mixture of different hardware
versions, variants and country-based modifications. Consequently,
it will be harder for the user to choose the correct image for
his device.

For the TL-WR941N/ND v3, so far one had to use the image of the
v2 revision. This patch creates a separate definition for v3, so
there is less room for confusion, as the correct version is
indicated by the image name.

Signed-off-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>

adrianschmutzler commented:
I've made a compromise: I think it is valid to have a separate definition for the 941v3, which a created in a second patch. Having the precise revision in the image name should actually be a relevant benefit to justify the additional image building process.
However, I do not think this is justifiable for the N/ND versions.

adrianschmutzler commented:
Rebased.

adrianschmutzler commented:
ALT patch is merged, so this is not blocked anymore.

adrianschmutzler commented:
@981213 Maybe have a look whether you want to merge this. It's not particularly complicated, but just lacks a decision on whether it's acceptable or not (and you have followed most of the discussion).

981213 commented:
I think it's unnecessary:
1. These are ancient 4/32 devices and it's unlikely that anyone would buy it nowadays. I think a sudden introduction of tl-wr941n-v3 may also be confusing for old users.
2. for wr941n/nd v4 since tp-link don't mind to call it without nd suffix in their factory firmware and again it's an ancient router, it isn't so important for us to do a split.

However this is just my personal opinions and I don't have a strong reason to reject this PR, so I decided to keep silent on this one :)

adrianschmutzler commented:
Since this one is not really important, since nobody cares about it, and for the reasons rightly given by 981213, I will close this PR.
